### PR TITLE
Webhook descriptors

### PR DESCRIPTION
**Pull request by tcmRyan**
Allow webhooks to be generated from registered webhooks

### CODE REVIEW

File: server/desciptor.py [hunk 1]
```diff
@@ -39,4 +40,18 @@ def __init__(self, key, name, description, vendor_name,
...
+        self.build_description()
+
+    def build_description(self):
+        self.descriptor['modules']['webhooks'] = self.webhooks
+        pass
+
+    @property
+    def webhooks(self):
+        webhooks = []
+        for webhook in WebHook:
```
Comment (jstnlef) [inline server/desciptor.py]: can be a comprehension

File: server/desciptor.py [hunk 2]
```diff
@@ -39,4 +40,18 @@ def __init__(self, key, name, description, vendor_name,
...
             },
-        }
\ No newline at end of file
+        }
+
+        self.build_description()
+
+    def build_description(self):
+        self.descriptor['modules']['webhooks'] = self.webhooks
+        pass
```
Comment (jstnlef) [inline server/desciptor.py]: The `pass` is extraneous

File: server/webhooks/__init__.py [hunk 3]
```diff
@@ -0,0 +1,60 @@
+from urllib.parse import urlparse
+from server.atl_requests import Context
+from flask import url_for
+
+
+class WebhookContext(Context):
```
Comment (jstnlef) [inline server/webhooks/__init__.py]: Not sure this needs to be its own class. Presumably you can just create a `Context` and move the `parse_callback` call to a standalone helper function in the module.

File: server/webhooks/__init__.py [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    def __iter__(cls):
+        return iter(cls.registry)
+
+    def __str__(cls):
+        if cls in cls.registry:
+            return cls.__name__
+        return cls.__name__ + ": " + ", ".join([sc.__name__ for sc in cls])
+
+
+class WebHook(object, metaclass=RegisterWebHooks):
```
Comment (jstnlef) [inline server/webhooks/__init__.py]: So presumably you will be able to do `WebHook.registry` to get the registered classes right?

File: server/webhooks/__init__.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+        cls.registry.add(cls)
+        cls.registry -= set(bases)
+
+    def __iter__(cls):
+        return iter(cls.registry)
+
+    def __str__(cls):
+        if cls in cls.registry:
+            return cls.__name__
+        return cls.__name__ + ": " + ", ".join([sc.__name__ for sc in cls])
```
Comment (jstnlef) [inline server/webhooks/__init__.py]: Don't need the `[]` here

**Review by jstnlef — COMMENTED** [review]
I guess I'm not entirely sure why this is useful but it seems fine to me. As much as I don't particularly love metaclasses, this use seems pretty tame.

File: server/webhooks/__init__.py (hunk 4 above)
Comment (tcmRyan) [reply server/webhooks/__init__.py]: correct, which is the question, does the metaclass approach make sense or should I have opted for a different pattern

**Comment by tcmRyan** [comment]
The intent here is so that everything is defined in one spot.  Right now you would create an endpoint and then need to go to the json descriptor and modify it.

**Comment by jstnlef** [comment]
I guess I'm still not quite understanding the problem. Seems as though both of your Callback classes have `descriptor` defined still?

**Comment by tcmRyan** [comment]
Right, the idea is that the descriptor gets defined in the callback.  If I don't to it that way it means that I have to maintain an entry in the master descriptor for each callback. Which means if I remove, modify, or add a callback I would have to remember to change the separate json descriptor file as well.

**Pull request merged**